### PR TITLE
Devicemapper storage

### PR DESCRIPTION
**Pull request by amshinde**
Check for devicemapper storage and use the underlying block device instead of overlay file system.

### CODE REVIEW

**Comment by amshinde** [comment]
Will be adding in more unit tests later.

File: src/util.c [hunk 1]
```diff
@@ -709,6 +714,86 @@ dup_over_stdio(int *fdp){
...
+		close(fd);
+		return NULL;
+	}
+
+	close(fd);
+	return buf;
+}
+
+// Legacy naming scheme for virtio block devices
+// Reference : https://github.com/torvalds/linux/blob/master/drivers/block/virtio_blk.c#L478
```
Comment (jodh-intel) [inline src/util.c]: Ah - I see that you've copied this verbatim (with a very minor tweak to appease a compiler warning?) Don't we atleast need a copyright attribution here for the FSF?

File: src/mount.c [hunk 2]
```diff
@@ -755,3 +755,68 @@ cc_is_devicemapper(uint major, uint minor)
...
+		g_warning("Absolute path not provided %s", path);
+		return NULL;
+	}
+
+	if (g_strcmp0(path, "/") == 0 ) {
+		return g_strdup(path);
+	}
+
+	if (stat(path, &cur_stat) == -1) {
+		return NULL;
```
Comment (grahamwhaley) [inline src/mount.c]: Should we have a g_warning for this one?

File: src/mount.c [hunk 3]
```diff
@@ -820,3 +820,56 @@ cc_mount_point_for_path(const gchar *path) {
...
+		g_debug("Invalid input for proc_mounts_path");
+		return ret;
+	}
+
+	/* note: glib api "g_unix_mounts_get" for fetching mount info fails in
+	 * case of bind mounts. Use getmntent_r instead.
+	 */
+
+	file = setmntent(PROC_MOUNTS_FILE, "r");
+	if ( !file) {
```
Comment (grahamwhaley) [inline src/mount.c]: bracket whitespace inconsistency?

File: src/oci.c [hunk 4]
```diff
@@ -229,6 +229,11 @@ cc_oci_get_config_and_state (gchar **config_file,
 			(*state)->procsock_path,
```
Comment (grahamwhaley) [inline src/oci.c]: nit: ' , ' whitespace in commit message, and this one is 'tests' and the previous one is 'test' :-)

File: src/util.c [hunk 5]
```diff
@@ -740,6 +745,55 @@ get_random_bytes(uint num) {
 	return buf;
 }
 
+// Legacy naming scheme for virtio block devices
```
Comment (grahamwhaley) [inline src/util.c]: I'll note, these are the only C++ style comments in this file ;-)

File: src/util.c [hunk 6]
```diff
@@ -740,6 +745,55 @@ get_random_bytes(uint num) {
...
+static int
+virtblk_name_format(char *prefix, int index, char *buf, int buflen)
+{
+	const int base = 'z' - 'a' + 1;
+	char *begin = buf + strlen(prefix);
+	char *end = buf + buflen;
+	char *p;
+	int unit;
+
+	p = end - 1;
```
Comment (grahamwhaley) [inline src/util.c]: well, there is a mini mind melting algo - if we'd written it, I'd be asking for an explanatory comment - but, as it is a clone of the 'standard', that's fine.

**Comment by chavafg** [comment]
qa-failed

**Comment by chavafg** [comment]
Got this functional test error: 

```
not ok 3 exec false
# (from function `verify_runtime_dirs' in file tests/functional/common.bash, line 161,
#  from function `testcontainer' in file tests/functional/common.bash, line 227,
#  in test file tests/functional/exec.bats, line 53)
#   `testcontainer "$container_id" "running"' failed
# Running: timeout 5 /home/cloud/go/src/github.com/01org/cc-oci-runtime/tests/functional/../../cc-oci-runtime  --debug --root /tmp/tmp.DexA6eLXsi-cor-test --global-log /run/cc-oci-runtime/cc-oci-runtime.log --hypervisor-log-dir /tmp/tmp.JwlAyuE8r2-cor-test --shim-path "/home/cloud/go/src/github.com/01org/cc-oci-runtime/cc-shim" --proxy-socket-path "/home/cloud/go/src/github.com/01org/cc-oci-runtime/tests/functional/proxy.sock" run -d --bundle /var/lib/oci/bundle tests_id
# this is a test
# README
# README.rst
# common.bash
# common.bash.in
# data
# exec.bats
# help.bats
# kill.bats
# pause.bats
# proxy.sock
# run-functional-tests.sh
# run-functional-tests.sh.in
# start.bats
# state.bats
# version.bats
# README
# README.rst
# common.bash
# common.bash.in
# data
# exec.bats
# help.bats
# kill.bats
# pause.bats
# proxy.sock
# run-functional-tests.sh
# run-functional-tests.sh.in
# start.bats
# state.bats
# version.bats
# exit code: 0
# expected exit code: 0
# Running: /home/cloud/go/src/github.com/01org/cc-oci-runtime/tests/functional/../../cc-oci-runtime  --debug --root /tmp/tmp.DexA6eLXsi-cor-test --global-log /run/cc-oci-runtime/cc-oci-runtime.log --hypervisor-log-dir /tmp/tmp.JwlAyuE8r2-cor-test --shim-path "/home/cloud/go/src/github.com/01org/cc-oci-runtime/cc-shim" --proxy-socket-path "/home/cloud/go/src/github.com/01org/cc-oci-runtime/tests/functional/proxy.sock" list --format table | grep tests_id | grep running
# exit code: 0
# expected exit code: 0
# SUCCESS
# Test logs: /tmp/tmp.DexA6eLXsi-cor-test
# Global log file: /run/cc-oci-runtime/cc-oci-runtime.log
# Hypervisor log dir: /tmp/tmp.JwlAyuE8r2-cor-test
# bats status: 0
# bats output (lines array contains 7 lines):
# lines[0]: 'console.sock'
# lines[1]: 'ga-ctl.sock'
# lines[2]: 'ga-tty.sock'
# lines[3]: 'hypervisor.sock'
# lines[4]: 'process.sock'
# lines[5]: 'state.json'
# lines[6]: 'workload'
# BUNDLE_DIR: /var/lib/oci/bundle
# ROOTFS_DIR: /var/lib/oci/bundle/rootfs
```

Seems like `lines[6]: 'workload'` is the issue: 
```
[ "${lines[6]}" = "" ] 
```
from https://github.com/01org/cc-oci-runtime/blob/master/tests/functional/common.bash.in#L161

**Comment by dlespiau** [comment]
This is quite a large change. You we open master to 2.2.x and branch out a 2.1.x stable branch?

**Comment by grahamwhaley** [comment]
@dlespiau ack from me - good idea on the version bump

**Comment by jcvenegas** [comment]
Running checkcommits version 0.0.1 (commit 2f18bd36aedcb97213dc794be2d5fb1306f9015e)
Detected TravisCI Environment
Found 21 commits between commit db9196bc532c6deebfabeaea4542fea254ea1821 and branch master
Checking commit 4e966691b70244ee4e1cb1236d7a639e7810b286
Checking commit e9ddabc927bfeaa4ec8ad91670b590f39ef79252
Checking commit 867e277b1fefa9395bb9adb72564cb75245b2644
Checking commit 93b13a3c816cdbf082a424e8099ad2fc94e7485c
Checking commit 56bc0929abbd91e86de77140830cfffb76e21310
Checking commit 33c8abac4ccccbd0f530d00473611f47e274fde9
Checking commit 00189f1bfff13436ba228c395b657fbaa8e76d56
Checking commit 2035ae61cd49b607aab46f92673c0f6f60fda56d
Checking commit 9b66cf2a03fb62531df778a7681b9cd04547f54b
Checking commit 9097610b691b30277cb0c9919af4b8bd9984ce58
Checking commit f97694e4763dd615e307d61007a226ac9eb621c6
Commit f97694e4763dd615e307d61007a226ac9eb621c6: single-line "Signed-off-by" body not permitted

File: src/mount.c [hunk 7]
```diff
@@ -309,6 +309,64 @@ cc_handle_mounts(struct cc_oci_config *config, GSList *mounts, gboolean volume)
...
+ *
+ * \return \c cc_oci_mount on success, else \c NULL.
+ */
+struct cc_oci_mount*
+rootfs_bind_mount(struct cc_oci_config *config)
+{
+	struct cc_oci_mount *m;
+
+	if (! config ) {
+		return false;
```
Comment (amshinde) [inline src/mount.c]: yes, this needs to be NULL. Fixing this.

**Comment by chavafg** [comment]
qa-failed

**Comment by chavafg** [comment]
qa-failed

File: src/proxy.c [hunk 8]
```diff
@@ -1274,6 +1274,40 @@ cc_proxy_hyper_pod_create (struct cc_oci_config *config)
...
+	JsonArray  *fsmap_arr;
+	JsonObject *fsmap_desc;
+	GSList     *l;
+
+	if (! config) {
+		return NULL;
+	}
+
+	fsmap_arr = json_array_new();
+	for (l = config->oci.mounts; l && l->data; l = g_slist_next (l)) {
```
Comment (sameo) [inline src/proxy.c]: So this is only handling volumes, or both volumes and rootfs ?

File: src/mount.c [hunk 9]
```diff
@@ -874,3 +880,62 @@ cc_get_device_and_fstype(gchar *mount_point, gchar **device_name, gchar **fstype
...
+       }
+
+       gchar *container_root = config->oci.root.path;
+
+       if (! cc_device_for_path(container_root, &major, &minor)) {
+               g_critical("Could not get the underlying device for rootfs");
+               return false;
+       }
+
+       if (! cc_is_devicemapper(major, minor)) {
```
Comment (sameo) [inline src/mount.c]: Could we make that one a little more generic, and have a `cc_is_blockdevice()` that could eventually also support other block device based graph drivers ?

File: src/mount.c [hunk 10]
```diff
@@ -874,3 +880,62 @@ cc_get_device_and_fstype(gchar *mount_point, gchar **device_name, gchar **fstype
...
+ * Check for devicemapper storage for container rootfs.
+ * If devicemapper block device is detected, set the
+ * config->device_name to the device found.
+ *
+ * \param config \ref cc_oci_config.
+ *
+ * \return \c true on success, else \c false.
+ */
+gboolean
+cc_oci_check_for_devicemapper(struct cc_oci_config *config)
```
Comment (sameo) [inline src/mount.c]: We could make that one a little more generic and call it e.g. `cc_oci_rootfs_is_block_device()`

File: src/util.c [hunk 11]
```diff
@@ -740,6 +745,65 @@ get_random_bytes(uint num) {
 	return buf;
 }
 
+// Legacy naming scheme for virtio block devices
+// Reference : https://github.com/torvalds/linux/blob/master/drivers/block/virtio_blk.c#L478
```
Comment (sameo) [inline src/util.c]: This specific file happens to not have any explicit copyright owners, so we should not add any attribution to our file.
For people to check who wrote this code, please replace that comment with:
```C
// Reference: linux/blob/master/drivers/block/virtio_blk.c @ <SHA1>
```
From there a git log will tell who wrote the code before we copied it. Additionally, potential future copyright owners will not be able to claim ownership of this code.

**Comment by sameo** [comment]
@dlespiau The version bump makes sense, this is a significant change. As for the branch, opening a stable branch means this is something we'll have to maintain. I'd rather not be doing this.

File: src/mount.c [hunk 12]
```diff
@@ -243,13 +245,46 @@ cc_handle_mounts(struct cc_oci_config *config, GSList *mounts, gboolean volume)
 			continue;
 		}
 
-		if ((! cc_pod_is_vm(config) || cc_pod_is_pod_sandbox(config)) && volume) {
+		if (! volume) {
+			/* bind mount container rootfs. We do this for regular
+			 * container as well for a pod sandbox.
+			 */
```
Comment (amshinde) [inline src/mount.c]: Will change that to include cc_pod_is_vm() check. Although this bind-mounting change with the added check doesnt seem to play well with crio. Still taking a look at this.

File: src/proxy.c (hunk 8 above)
Comment (amshinde) [inline src/proxy.c]: Just the volumes. The oci.mounts should contain only volumes right? config->rootfs_mounts should contain just the rootfs mounts if I have understood correctly.

Comment (sameo) [inline src/proxy.c]: For pods, that's correct. I see you changed that logic with that PR, and that sounds reasonable to me.

File: src/util.c [hunk 13]
```diff
@@ -709,6 +709,37 @@ dup_over_stdio(int *fdp){
...
 
+/**
+ * Generated random bytes.
+ *
+ * \param size Number of bytes to be generated.
+ *
+ * \return \c Newly allocated buffer on success, NULL otherwise.
+ **/
+uint8_t *
+get_random_bytes(uint num) {
```
Comment (amshinde) [inline src/util.c]: @sameo Seems like this is  new system call introduced in 3.17, would not be available in older kernels. The system call itself was added very recently in glibc: https://lwn.net/Articles/711013/.
I think reading from /dev/urandom is fairly standard way to get random data.

**Comment by amshinde** [comment]
@sameo I have addressed your other review comments.

**Comment by sameo** [comment]
LGTM

**Comment by chavafg** [comment]
qa-failed

**Comment by chavafg** [comment]
qa-failed

**Comment by chavafg** [comment]
qa-passed

**Comment by jcvenegas** [comment]
@jodh-intel  could you review this PR again ?

File: src/util.c [hunk 14]
```diff
@@ -709,6 +714,96 @@ dup_over_stdio(int *fdp){
...
+	int fd;
+	char *file = "/dev/urandom";
+	uint8_t *buf;
+
+	buf = g_malloc0(num);
+
+	fd = open(file, O_RDONLY);
+	if (fd < 0) {
+		g_critical("Error opening %s: %s", file, strerror(errno));
+		return NULL;
```
Comment (jodh-intel) [inline src/util.c]: This leaks `buf`.

File: src/mount.c [hunk 15]
```diff
@@ -309,6 +351,91 @@ cc_handle_mounts(struct cc_oci_config *config, GSList *mounts, gboolean volume)
...
+		goto error;
+	}
+
+	/* Type */
+	m->mnt.mnt_type = g_strdup("bind");
+	if (! m->mnt.mnt_type) {
+		goto error;
+	}
+
+	g_debug("Rootfs bind mounting done");
```
Comment (jodh-intel) [inline src/mount.c]: ? This function doesn't call mount.

Can you add a `g_debug()` call somewhere showing details of this mount (as we do for the other mounts - see `cc_handle_mounts()`).

File: src/util.c [hunk 16]
```diff
@@ -709,6 +714,86 @@ dup_over_stdio(int *fdp){
...
+	}
+
+	close(fd);
+	return buf;
+}
+
+// Legacy naming scheme for virtio block devices
+// Reference : https://github.com/torvalds/linux/blob/master/drivers/block/virtio_blk.c#L478
+static int
+virtblk_name_format(char *prefix, int index, char *buf, int buflen)
```
Comment (amshinde) [inline src/util.c]: I have added checks for these

File: src/util.c (hunk 1 above)
Comment (amshinde) [inline src/util.c]: @As per the discussion on this, I have updated the reference to include the git sha

File: src/mount.c (hunk 15 above)
Comment (amshinde) [inline src/mount.c]: @jodh-intel I'll change the message here to "Added rootfs bind mount". The details of the rootfs mount should be displayed by cc_handle_mounts

Comment (jodh-intel) [inline src/mount.c]: I still think that's a little unclear - that function isn't doing anything with mounts - it's just allocating a struct/object so I'd just log something like:

```
Created rootfs bind mount object
```

I'm not really sure even that is useful though, unless you add some details of the object itself maybe?

Comment (amshinde) [inline src/mount.c]: How about  
g_debug("Created rootfs bind mount for container %s", config->optarg_container_id)

Comment (jodh-intel) [inline src/mount.c]: `s/bind mount for/bind mount object for/` wfm - I just think it's too misleading to suggest we've created a bind mount unless we've actually called `mount(2)`.

File: src/util.c (hunk 14 above)
Comment (amshinde) [inline src/util.c]: fixed

File: src/mount.c (hunk 15 above)
Comment (amshinde) [inline src/mount.c]: I have pushed with changed comment as bind mount object.

**Comment by chavafg** [comment]
qa-passed

**Comment by jcvenegas** [comment]
just tested with hyperstart changes, tested that now dnf works with devicemapper, after the image is updated this.

**Comment by chavafg** [comment]
qa-passed

**Comment by chavafg** [comment]
I have tested this on Ubuntu, Fedora, Centos and ClearLinux with both overlay and devicemapper as the docker storage driver and no regressions found. 

lgtm

**Pull request merged**